Continuous inter-point constraints

AVHopp opened the pull request:
This PR introduces a first variant of inter-point constraints by using the `botorch`-provided interface.

Here, an inter-point constraint is a constraint that acts on a full batch instead of a single recommendation. If we think of a batch recommendation of a matrix with shape `batches x features`, then our previous constraints would be row-wise, while these constraints allow mixed constraints across both dimensions.

This PR introduces new classes, and these classes were modeled similar to the already existing classes for continuous constraints. Some things had to be changed though resp. we need to aware:
1. Most importantly, the way of sampling points from a polytope. The reason is that `get_polytope_samples` is not made for including inter-point constraints, and hence a workaround was implemented. The workaround basically transforms the space in a one-dimensional space with `batches * features` many features, and then defines both normal and inter-point constraints over this space (see https://github.com/pytorch/botorch/issues/2468). Note that this might interfere with #301 and that some alignment might be necessary.
2. The constraints in this PR always combines full columns. That is, the constraints $x_1 <=1$ would be interpreted as "the sum of $x_1$ across the whole batch needs to be smaller than 1". This can e.g. be used to limit the usage of resources, like only having 100ml of a substance available for the full batch. Also, different columns can be combined, so having $x_2 + 2*x_3 <=100$ would mean "sum of $x_1$ plus two times the sum of $x_3$ across the batch has to be smaller than 100$.
3. For now, the constraints have only been applied for continuous search spaces. The reason for this is that I want to keep this PR rather short and use it to discuss in general about the design, and to avoid any potential conflicts with #301 . However, including them in hybrid search spaces should be possible already in this PR. For discrete spaces, this still needs some investigation. So interpret this as just a first step in the direction of more :)

AVHopp commented:
Note: This PR is currently on hold for two reasons:
1. There was a misunderstanding between @AdrianSosic and me what "inter-point" constraint actually means, and incorporating the necessary changes might require some more redesign.
2. There is another open PR which currently changes the behavior of `get_polytope_samples`, and that PR should be merged first.

AVHopp commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
```
@Scienfitz @AdrianSosic this PR is now ready to review, as Martin and I discussed that this can already be included independent of different kinds of interpoint constraints we might introduce later.

AVHopp commented on CHANGELOG.md:
TODO: Check for Hybrid spaces

Scienfitz commented:
as mentioned in our call but posting here for documentation / reference: I think its worth exploring whether a design not via deriving from `ContinuousLinearConstraint` but by adding a simple switching flag to it might be better

AdrianSosic reviewed (changes requested):
Not yet completely done with the review, but because time is running, here a first batch  of comments

Scienfitz commented on baybe/searchspace/continuous.py:
```diff
@@ -383,14 +408,17 @@ def sample_uniform(self, batch_size: int = 1) -> pd.DataFrame:
             return self._sample_from_bounds(batch_size, self.comp_rep_bounds.values)
 
+        if self.has_interpoint_constraints:
+            return self._sample_from_polytope_with_interpoint_constraints(
```
is it feasible to  keep splittitng the sample function up like that? eg what happens if we have interpoint+cardinality constraints

Scienfitz replied:
what do you mean exactly?
Is my udnerstadning right that sample from polytope just arranges the cosntraint types accordingly and there are no multiple sample from polytope functions remaining?

AVHopp replied:
Exactly, the two functions have been merged, so there is only a single `_sample_from_polytope` function left. This will be called when there are no cardinality constraints

AVHopp replied:
Then it would call the `_sample_from_polytope_with_cardinality_constraints` function and interpoint constraints would indeed be ignored. But looking at the code, don't normal constraints also get ignored when there are cardinality constraints?

Scienfitz replied:
maybe @AdrianSosic can clarify (in a not too long text :) why it is that way

can be accepted if this is intended or required temporarily or as workaround, just making sure this is not accidental

AdrianSosic replied:
@AVHopp: I think the reasoning is not perfectly accurate here. The current logic is:
* At the moment, the only existing nonlinear constraint is the `ContinuousCardinalityConstraint`. Its sampling works by sampling active/inactive sets and then sampling from a polytope on the active set.
* Linear constraints are taken into account by polytope sampling, or to put it differently: polytope sampling IS sampling with linear constraints.

--> Bottom line: all constraints are respected. And we should make sure that stays also true for the interpoint ones. 

Helpful?

AVHopp replied:
Resolving here since interpoint constraints and cardinality constraints can now not be used together since this is what was agreed upon internally.

Scienfitz commented on baybe/searchspace/continuous.py:
```diff
@@ -383,14 +408,17 @@ def sample_uniform(self, batch_size: int = 1) -> pd.DataFrame:
             return self._sample_from_bounds(batch_size, self.comp_rep_bounds.values)
 
+        if self.has_interpoint_constraints:
+            return self._sample_from_polytope_with_interpoint_constraints(
```
isnt 'sample_from_polytope' a special case of the new function? If so I think it could be better design if one was contained int he other or one calls the other

AVHopp replied:
You are right, I have combined the two functions into a single one and will resolve this comment here once the changes have been pushed- I will keep the other one open for further discussion of the topic.

Scienfitz commented on CHANGELOG.md:
the new intperpoint flag should be mentioned in the userguide

Scienfitz replied:
ok ty

I think its missing a sentence about a praticaly relevante xample where iterpoint cosntriants would be sueful

Also, is the four 3 level common to sue? I havent looked at the entire file but putting it at the `####` level seems strange to me

AVHopp replied:
Was thinking that it is a subpoint of `ContinuousLinearConstraints`, but I now removed that layer again. Also added a sentence on the motivation, feel free to resolve if this is suitable.

Scienfitz replied:
ty

AdrianSosic commented on baybe/constraints/continuous.py:
```diff
@@ -45,6 +46,15 @@ class ContinuousLinearConstraint(ContinuousConstraint):
     rhs: float = field(default=0.0, converter=float, validator=finite_float)
     """Right-hand side value of the in-/equality."""
 
+    is_interpoint: bool = field(default=False)
```
we could also do it in the smart way a la `is_interpoint: bool = field(alias="interpoint", default=True)` which would give us best of both worlds?

AVHopp replied:
I'd say let's go easy here, I'll replace by `interpoint`.

AVHopp commented on CHANGELOG.md:
Will stick with `interpoint` then since function names like `has_interpoint_constraints` look better than `has_inter-point_constraints` or `has_inter_point_constraints`

AVHopp replied:
Did a search-and-replace. If you spot any point where this is still incorrect, please resolve here.

AVHopp commented on tests/conftest.py:
```diff
@@ -895,7 +923,10 @@ def fixture_default_onnx_surrogate(onnx_str) -> CustomONNXSurrogate:
+    campaign: Campaign,
+    n_iterations: int,
+    batch_size: int,
+    add_noise: bool = True,
```
Seems to be a legit reformat, at least I cannot change it back without ruff immediately complaining, so I'd leave it as-is now. Please resolve if that is ok

Scienfitz replied:
hmm, theres several such reformats in this PR that might or might note be intended. Please make sure your suing the right verisons and envs and if thats the case resolve this

AVHopp replied:
I double checked and everything seems to be fine:
![image](https://github.com/user-attachments/assets/50ec3daa-6253-4555-ba17-f497b415a4ce)
Hence resolving.

AVHopp commented on tests/constraints/test_constraints_continuous.py:
```diff
@@ -68,6 +68,81 @@ def test_inequality3(campaign, n_iterations, batch_size):
 
+@pytest.mark.parametrize("parameter_names", [["Conti_finite1", "Conti_finite2"]])
+@pytest.mark.parametrize("constraint_names", [["InterConstraint_1"]])
+@pytest.mark.parametrize("batch_size", [5], ids=["b5"])
```
I'll just remove the batch_sizes for the test. Would prefer to keep different numbers of iterations since it previously happened that it seemed like constraints were obeyed, while this was just a numerical coincidence, and this is at least a bit improved when doing more iterations. Please resolve if this is ok.

Scienfitz replied:
ok

AdrianSosic commented on baybe/searchspace/continuous.py:
```diff
@@ -282,6 +286,24 @@ def comp_rep_bounds(self) -> pd.DataFrame:
+        return any(
+            c.is_interpoint for c in self.constraints_lin_eq + self.constraints_lin_ineq
+        )
+
```
Yes, I'm aware 👍🏼 Just for the newly introduced ones

AVHopp commented on baybe/constraints/continuous.py:
```diff
@@ -98,7 +108,10 @@ def _drop_parameters(
+        self,
+        parameters: Sequence[NumericalContinuousParameter],
+        idx_offset: int = 0,
+        batch_size: int = 1,
```
Removed the default

AVHopp commented on examples/Constraints_Continuous/linear_constraints.py:
```diff
@@ -140,3 +142,59 @@
 )
+
+
+### Using inter-point constraints
```
Created a new example, please have a look and resolve if happy

Scienfitz replied:
ty

AVHopp commented on examples/Constraints_Continuous/linear_constraints.py:
```diff
@@ -140,3 +142,59 @@
+# Since these constraints require information about the batch size, they are not used
+# during the creation of the search space but handed over to the `recommend` call.
+# This example models the following inter-point constraints and combines them also
+# with regular constraints.
```
Was deleted with the change of the example.

AVHopp commented on examples/Constraints_Continuous/linear_constraints.py:
```diff
@@ -140,3 +142,59 @@
+    rec["Target"] = target_values
+    inter_campaign.add_measurements(rec)
+    # Check inter-point constraints
+    assert rec["x_1"].sum() >= 2.5 - TOLERANCE
```
Changed from `assert` to `print` such that it is more in-line with the existing example. If `assert` should be reintroduced, just ping me here and unresolve.

AVHopp commented on tests/constraints/test_constraints_continuous.py:
```diff
@@ -68,6 +68,81 @@ def test_inequality3(campaign, n_iterations, batch_size):
+@pytest.mark.parametrize("parameter_names", [["Conti_finite1", "Conti_finite2"]])
+@pytest.mark.parametrize("constraint_names", [["InterConstraint_4"]])
+@pytest.mark.parametrize("batch_size", [5], ids=["b5"])
+def test_interpoint_inequality_multiple_parameters(campaign, n_iterations, batch_size):
```
Well, I followed the structure that we already had, where there is a dedicated test for each individual constraint. I would thus stay with the current design since it is more in line with the current design and better readable in my opinion.

Scienfitz replied:
not sure where thats consistent, dont they all test the same constaint just a different configuration? thats pretty much a use case for parameterization

AVHopp replied:
"Consistent" in the sense of "there is a dedicated test per constraint". Also, we do the very same thing for other constraints, e.g. here:
```python
@pytest.mark.parametrize("parameter_names", [["Conti_finite1", "Conti_finite2"]])
@pytest.mark.parametrize("constraint_names", [["ContiConstraint_3"]])
@pytest.mark.parametrize("batch_size", [5], ids=["b5"])
def test_inequality1(campaign, n_iterations, batch_size):
    """Test inequality constraint with equal weights."""
    run_iterations(campaign, n_iterations, batch_size, add_noise=False)
    res = campaign.measurements
    print(res)
    assert (1.0 * res["Conti_finite1"] + 1.0 * res["Conti_finite2"]).ge(0.299).all()
@pytest.mark.parametrize("parameter_names", [["Conti_finite1", "Conti_finite2"]])
@pytest.mark.parametrize("constraint_names", [["ContiConstraint_4"]])
@pytest.mark.parametrize("batch_size", [5], ids=["b5"])
def test_inequality2(campaign, n_iterations, batch_size):
    """Test inequality constraint with unequal weights."""
    run_iterations(campaign, n_iterations, batch_size, add_noise=False)
    res = campaign.measurements
    print(res)
    assert (1.0 * res["Conti_finite1"] + 3.0 * res["Conti_finite2"]).ge(0.299).all()
```
Also, do I understand it correctly that you would then prefer to have the code then in the style of 
```python

    for batch in range(n_iterations):
        res_batch = res[res["BatchNr"] == batch + 1]
        assert campaign.searchspace.constraints[0].coefficients[0] * res_batch["Conti_finite1"].sum() >= campaign.searchspace.constraints[0].rhs
```
or similar? In my opinion, that is significantly harder to read, so I am still not convinced that the design here should change.

Scienfitz replied:
ok fine, my suggestion was coming form the persepctive of reducing code (whihc i think is possible here by a factor of 3 to 4 or so) but I guess it would reduce readability as you said

AVHopp commented on tests/conftest.py:
```diff
@@ -553,6 +553,34 @@ def custom_function(df: pd.DataFrame) -> pd.Series:
+        "InterConstraint_4": ContinuousLinearConstraint(
+            parameters=["Conti_finite1", "Conti_finite2"],
+            coefficients=[2, -1],
+            operator=">=",
```
Included a corresponding constraint (which by the way showed that there was a significant error, so thanks for pointing this out!)

Scienfitz commented on examples/Constraints_Continuous/interpoint.py:
```diff
@@ -0,0 +1,119 @@
+# interpoint constraints do so **across** the points of the batch. That is, an
+# interpoint constraint of the form ``x_1 + x_2 <= 1`` enforces that the sum of all
+# ``x_1`` values plus the sum of all ``x_2`` values in the batch must not exceed 1.
+
```
feels like this misses a sentence giving a concrete example where this could be useful in the real world.

AVHopp replied:
Added a sentence, please resolve ifg happy.

Scienfitz commented on tests/constraints/test_constraints_continuous.py:
```diff
@@ -68,6 +68,96 @@ def test_inequality3(campaign, n_iterations, batch_size):
+        print(campaign.searchspace.constraints)
+        assert (
+            2 * res_batch["Conti_finite1"].sum() - res_batch["Conti_finite2"].sum()
+            <= 0.301
```
I'd prefer if tolerance aspects are not hardcoded into the number you write here but in a constant defined at the top of the file and expressions just us `0.3 + TOLERANCE` etc
it has been done in other test files I think

AVHopp replied:
Done, also fixed an incorrect test that I spotted in the tests.

Scienfitz commented on baybe/constraints/continuous.py:
```diff
@@ -117,16 +134,29 @@ def to_botorch(
+            coefficients = self.coefficients
+            torch_indices = torch.tensor(param_indices)
+        else:
+            assert batch_size is not None
```
```suggestion
            assert batch_size is not None, "Message explaining the problem"
```

AVHopp replied:
Added an error message.

AVHopp commented:
@AdrianSosic @Scienfitz this PR is now (once again) ready for review. As agreed, I added a validation preventing the use of interpoint and cardinality constraints together. Also, I cleaned the gir history a little bit and rebase onto `main`, so this should be as clean as possible for you to review :)

Scienfitz commented on baybe/constraints/continuous.py:
```diff
@@ -117,16 +134,31 @@ def to_botorch(
+        else:
+            assert (
+                batch_size is not None
+            ), "No batch_size set but using interpoint constraints"
```
dont really like this error message, think of a suer getting this. they dont set anything as this is internal so this will confuse

AVHopp replied:
So can you propose an alternative then? Could add something like "This should never happen and means that there is a bug in the code" to make sure that users do not get confused by this?

Scienfitz replied:
yes it sounds like a `RuntimeError` to me

AdrianSosic replied:
Plus: note that `assert` is NOT meant to be used for input validation, i.e. there is a conceptual difference between asserts and error throwing. The former can even be disabled via a python flag!

AVHopp replied:
Replaced the `assert` by now raising an error. I'll resolve, if you want another error message just un-resolve.

Scienfitz commented on baybe/constraints/validation.py:
```diff
@@ -98,3 +101,25 @@ def validate_cardinality_constraints_are_nonoverlapping(
             )
+
+
+def validate_no_interpoint_and_cardinality_constraints(
```
why is this not part of the function `validate_constraints` above? see eg line 39

AVHopp replied:
I interpret this comment as "Why is this function not being called withing `validate_constraints`?" and not as "Why did you write an extra function for this and not simply include the code in `validate_constraints`?".
In this case, the answer is "because I missed that part, will add and resolve here once done".

AVHopp replied:
Changed and fixed.

Scienfitz commented on baybe/recommenders/pure/bayesian/botorch.py:
```diff
@@ -234,6 +234,11 @@ def _recommend_hybrid(
         """
+        if searchspace.continuous.has_interpoint_constraints:
+            raise NotImplementedError(
+                "Interpoint constraints are not available in hybrid spaces."
```
this is a big downside, is this remotely possible?

AVHopp replied:
I did not test and check whether this might be possible to keep the current PR as minimal as possible by first introducing interpoint constraints for purely continuous spaces. I will however investigate if it is easy to include, might require more investigation or seems to be impossible.

AVHopp replied:
Summoning @AdrianSosic as this might also be interesting for him.

I tried around a bit and I do not think that there is a conceptual problem. Good news: We should be able to include interpoint constraints also for hybrid spaces. Bad news: This will probably require yet another change and adjustment to the `get_polytope_sample` method and fiddling around with indices. For that reason, I would propose to *not* implement this now in this PR since this is aleady open for quite a long time and there are already potential conflicts between this PR and #301 - and the more stuff I change in the sampling, the harder bringing things together will be. Hence, I would propose to investigate this after this PR which is then "Interpoint constraints for purely continuous spaces" and #301 have been merged.

Everyone agreeing?

Scienfitz replied:
ok out of PR scope

AdrianSosic replied:
Yes, sounds good to me, let's try to close this PR as quickly as possible and then rather have a fresh start with an isolated new feature 👍🏼

Scienfitz commented on baybe/searchspace/continuous.py:
```diff
@@ -391,14 +412,12 @@ def sample_uniform(self, batch_size: int = 1) -> pd.DataFrame:
+        if not self.is_constrained:
             return self._sample_from_bounds(batch_size, self.comp_rep_bounds.values)
 
+        # If there are neither cardinality nor interpoint constraints, we sample
```
a bit confused by this comment, wehre is the check here done to make sure theres no interpoint?

AVHopp replied:
Good catch, this is a leftover comment from back when there were two sampling methods, one for interpoint and one for no-interpoint cases. In the current version, `_sample_from_polytope` just handles this internally, so I will simply adjust this comment.

Scienfitz commented on baybe/searchspace/continuous.py:
```diff
@@ -413,22 +432,71 @@ def _sample_from_bounds(self, batch_size: int, bounds: np.ndarray) -> pd.DataFra
+
+        eq_constraints, ineq_constraints = [], []
+
+        # We start with the general constraints before going to interpoint constraints
```
this comment implies there is a loop over the former, then over the latter, but in reality its one loop combined so its a bit confusing

AVHopp replied:
Yes, seems to be an outdated comment. Will remove.

Scienfitz commented on baybe/searchspace/continuous.py:
```diff
@@ -413,22 +432,71 @@ def _sample_from_bounds(self, batch_size: int, bounds: np.ndarray) -> pd.DataFra
         from botorch.utils.sampling import get_polytope_samples
 
+        # The number of parameters is needed at some places for adjusting indices
+        num_of_params = len(self.parameters)
```
shouldnt it better be based on columns in the comp rep? might be identical at the moment but basing it on the comp rep columns is future proof

AVHopp replied:
True that, will change.

Scienfitz commented on docs/userguide/constraints.md:
```diff
@@ -79,6 +79,37 @@ ContinuousLinearConstraint(
 A more detailed example can be found
 [here](../../examples/Constraints_Continuous/linear_constraints).
 
+### Interpoint constraints
```
please write all headings in telegraph style, most importantly all main word capitalized

I know its inconsistent in the code, but Adrian and I both use this, so lets make it consistent

AVHopp replied:
That's a long comment for "Typo c -> C" ;) Will change.

AVHopp replied:
I have changed this in the header of the section and of the admonition. Were these all places or did I miss some?

Scienfitz replied:
ok it seems you also agree with using telegraph style everywhere then (that wasnt clear to me)
theres currently many places where we use it inconsistenlty in examples and UG, but for this PR dont worry about it

AVHopp replied:
Yes, I definitely agree. Then let me resolve here and put a note in the issue for collecting stuff to fix in the documentation.

Scienfitz commented on docs/userguide/constraints.md:
```diff
@@ -79,6 +79,37 @@ ContinuousLinearConstraint(
+ContinuousLinearConstraint(
+    parameters=["x_1", "x_2"],
+    operator="<=",
+    coefficients=[1.0, 1.0],
```
why make this a meaningless anon examaple and not execute the solvent one you describe in words?

AVHopp replied:
What do you mean here exactly? Would you prefer to have the code also include stuff like actual `SubstanceParameter` and so on? My goal was to have this user guide as minimal as possible, by only showing what is described, and that is the definition of an interpoint constraint.

Scienfitz replied:
this code snipped doesnt use any actual parameters, just names
So my suggestion is to craft the names and numbers according to the solvent example you describe in words in the adjacent text, instead of these anonymous x_1, x_2 names

AVHopp replied:
Changed the code snippet, please resolve if happy.

Scienfitz replied:
many thx
I would still change `SolventUsed` to `SolventAmount` which is more typical for a parameter name (`used` could also indicate binary yes/no)

AVHopp replied:
Changed it, hence resolving.

Scienfitz commented on examples/Constraints_Continuous/interpoint.py:
```diff
@@ -0,0 +1,122 @@
+    )
+    print(
+        "The sum of `x_2` across all batches is exactly 5",
+        np.isclose(rec["x_2"].sum(), 5),
```
should also work the tolerance in that a la sum -> minus target -> abs -> lt(TOLERANCE)

Scienfitz commented on examples/Constraints_Continuous/interpoint.py:
```diff
@@ -0,0 +1,122 @@
+        np.isclose(rec["x_2"].sum(), 5),
+    )
+    print(
+        "The sum of `2*x_3` minus the sum of `x_4` across all batches is at least >= 5",
+        2 * rec["x_3"].sum() - rec["x_4"].sum() >= 2.5 - TOLERANCE,
```
target numbers dont seem to match

AVHopp replied:
Fixed.

Scienfitz commented on tests/test_searchspace.py:
```diff
@@ -268,3 +268,30 @@ def test_cardinality_constraints_with_overlapping_parameters():
+
+def test_cardinality_and_interpoint_constraints():
+    """Using cardinality and interpoint constraints together raises an error."""
+    parameters = (
+        NumericalContinuousParameter("c1", (0, 1)),
+        NumericalContinuousParameter("c2", (0, 1)),
+    )
```
why not request them via fixture?

AVHopp replied:
Reason was to keep the style of the tests in that file: All of them but the memory estimate explicitly create parameters and do not request them via fixtures, and I prefer to keep the same style here. Can however also change if you insist.

Scienfitz commented on tests/conftest.py:
Hypothesis tests for linar constraints needs to be updated

AVHopp replied:
Updated it by adding a corresponding boolean that is being drawn. As I am not 100% sure if this small change is enough, please have a quick look and resolve if happy :)

Scienfitz replied:
ty
seems to me for the current situation its fine, but certainly when large scale hypothesis objects are built together there need to be some valdiations because I guess many things will blow up if we allow a 50% chance of it being an interpoint right? or instance, if you built a campaign now and request linear constraints it blows up for a hybrid space. But the needed design implications can only be decided at that point in time so I guess its fine for now

AVHopp replied:
Let me add a comment to the respective part of the test such that we are aware of this.

AVHopp replied:
Done that, hence resolve here.

Scienfitz commented on tests/constraints/test_constraints_continuous.py:
```diff
@@ -68,6 +70,103 @@ def test_inequality3(campaign, n_iterations, batch_size):
+
+    for batch in range(n_iterations):
+        res_batch = res[res["BatchNr"] == batch + 1]
+        assert np.isclose(
```
to include our own tolerance you can et `atol` to `TOLERANCE` in `isclose`
alternatively chain pandas sum -> minus target -> abs -> lt(TOLERANE)

AVHopp replied:
Will add `atol` and then resolve here.

Scienfitz commented on tests/constraints/test_constraints_continuous.py:
```diff
@@ -68,6 +70,103 @@ def test_inequality3(campaign, n_iterations, batch_size):
+        print(campaign.searchspace.constraints)
+        assert (
+            2 * res_batch["Conti_finite1"].sum() - res_batch["Conti_finite2"].sum()
+            <= 0.3 + TOLERANCE
```
usage of pandas chaining seems to be a bit inconsistent int he tests, here you stop the chain while below you do the ineq checks also in the chain

AVHopp replied:
Can you elaborate a bit? I do not understand what you mean by "pandas chaining" and also do not see the inconsistency. Do you mean the somehow weird extractiong of `res_batch`?

Scienfitz replied:
this ises a pandas chaining:
![image](https://github.com/user-attachments/assets/a97a8109-1b0d-4510-a03f-3cf140233382)


this doenst use pandas chaining:
![image](https://github.com/user-attachments/assets/d80b947e-bb3b-466c-9cc0-a67a4be0b738)

and in the test file both variants are used

Scienfitz replied:
what about line 94

AVHopp replied:
True, I missed it in two tests. Have included it (now hopefully everywhere). Resolving, if you spot another place where I missed it just un-resolve.

Scienfitz reviewed (commented):
I think hypothesis are missing for the ew constraints

AdrianSosic commented:
@AVHopp, wie man in den Wald schreit 😄: anything happening here?

AVHopp commented:
This PR is outdated and a new version will probably be worked on.
